tests/relay/test_subscribe.py

```diff
@@ -0,0 +1,110 @@
+
+    def test_subscribe_with_multiple_overlapping_pubsub_topics(self):
+        self.ensure_subscriptions_on_nodes(self.main_nodes, VALID_PUBSUB_TOPICS[:3])
+        self.ensure_subscriptions_on_nodes(self.main_nodes, VALID_PUBSUB_TOPICS[1:4])
+        for pubsub_topic in VALID_PUBSUB_TOPICS[:4]:
+            self.wait_for_published_message_to_reach_peer(pubsub_topic=pubsub_topic)
+
+    def test_subscribe_with_empty_pubsub_topic_list(self):
+        self.ensure_subscriptions_on_nodes(self.main_nodes, [])
+
+    def test_subscribe_with_invalid_pubsub_topic_format(self):
+        success_pubsub_topics = []
+        for pubsub_topic in INVALID_PUBSUB_TOPICS:
+            logger.debug(f"Running test with payload {pubsub_topic}")
+            try:
+                self.ensure_subscriptions_on_nodes(self.main_nodes, pubsub_topic)
+                success_pubsub_topics.append(pubsub_topic)
+            except Exception as ex:
+                assert "Bad Request" in str(ex)
+        assert not success_pubsub_topics, f"Invalid Pubsub Topics that didn't failed: {success_pubsub_topics}"
+
+    def test_unsubscribe_from_single_pubsub_topic(self):
+        self.ensure_subscriptions_on_nodes(self.main_nodes, [self.test_pubsub_topic])
+        self.wait_for_published_message_to_reach_peer()
+        self.delete_subscriptions_on_nodes(self.main_nodes, [self.test_pubsub_topic])
+        self.check_publish_without_subscription(self.test_pubsub_topic)
+
+    def test_unsubscribe_from_all_pubsub_topics(self):
+        self.ensure_subscriptions_on_nodes(self.main_nodes, VALID_PUBSUB_TOPICS)
+        for pubsub_topic in VALID_PUBSUB_TOPICS:
+            self.wait_for_published_message_to_reach_peer(pubsub_topic=pubsub_topic)
+        self.delete_subscriptions_on_nodes(self.main_nodes, VALID_PUBSUB_TOPICS)
+        for pubsub_topic in VALID_PUBSUB_TOPICS:
+            self.check_publish_without_subscription(pubsub_topic)
+
+    def test_unsubscribe_from_sume_pubsub_topics(self):
+        self.ensure_subscriptions_on_nodes(self.main_nodes, VALID_PUBSUB_TOPICS)
+        for pubsub_topic in VALID_PUBSUB_TOPICS:
+            self.wait_for_published_message_to_reach_peer(pubsub_topic=pubsub_topic)
+        self.delete_subscriptions_on_nodes(self.main_nodes, VALID_PUBSUB_TOPICS[:3])
+        for pubsub_topic in VALID_PUBSUB_TOPICS[:3]:
+            self.check_publish_without_subscription(pubsub_topic)
+        for pubsub_topic in VALID_PUBSUB_TOPICS[3:]:
+            self.check_published_message_reaches_peer(pubsub_topic=pubsub_topic)
+
+    def test_unsubscribe_from_non_existing_pubsub_topic(self):
+        self.ensure_subscriptions_on_nodes(self.main_nodes, [self.test_pubsub_topic])
+        self.wait_for_published_message_to_reach_peer()
+        try:
+            self.delete_subscriptions_on_nodes(self.main_nodes, ["/waku/2/rs/999/99"])
+            if self.node1.is_nwaku():
+                pass  # nwaku doesn't fail in this case
```

Comment:
Am I right when I say that you mentioned you were gathering all differences between Waku implementations? Where does that live?

Reply:
Here's the [page](https://www.notion.so/Nwaku-vs-Gowaku-vs-Jswaku-diffs-b3e0e8f1e6cd4c6d9855b0c3c4634bc5) I started, it's not much but still better than nothing :)